File: Big Data/Using Python/HBase - Store and Retrieve time-series IoT sensor data.py
HBase Setup
Bash
# Start HBase
start-hbase.sh

# Start Thrift server (required for Python connection)
hbase thrift start

# Start REST server (required for R connection)
hbase rest start -p 8080

# Verify HBase is running
hbase shell
> status

# Required libraries
# pip install happybase
import happybase
import time
import random
import struct
from datetime import datetime, timedelta

# ===== Step 1: Connect to HBase =====
connection = happybase.Connection('localhost', port=9090)
connection.open()
print("Connected to HBase successfully!")
print("Available tables:", connection.tables())

# ===== Step 2: Create Table for IoT Sensor Data =====
table_name = 'iot_sensor_data'

# Delete table if exists (for fresh start)
if table_name.encode() in connection.tables():
    connection.disable_table(table_name)
    connection.delete_table(table_name)
    print(f"Existing table '{table_name}' deleted.")

# Create table with column families
connection.create_table(
    table_name,
    {
        'sensor': dict(),       # sensor metadata (id, type, location)
        'reading': dict(),      # actual readings (temperature, humidity, pressure)
        'status': dict()        # device status (battery, signal, alert)
    }
)
print(f"Table '{table_name}' created with column families: sensor, reading, status")

# ===== Step 3: Generate and Store IoT Sensor Data =====
table = connection.table(table_name)

# Sensor configurations
sensors = [
    {'id': 'SENSOR_001', 'type': 'temperature', 'location': 'Building_A_Floor1'},
    {'id': 'SENSOR_002', 'type': 'humidity', 'location': 'Building_A_Floor2'},
    {'id': 'SENSOR_003', 'type': 'pressure', 'location': 'Building_B_Floor1'},
    {'id': 'SENSOR_004', 'type': 'temperature', 'location': 'Building_B_Floor2'},
    {'id': 'SENSOR_005', 'type': 'humidity', 'location': 'Building_C_Floor1'},
]

# Generate time-series data (every 5 minutes for 24 hours = 288 readings per sensor)
random.seed(42)
base_time = datetime(2024, 6, 15, 0, 0, 0)
total_records = 0

print("\nStoring IoT sensor data...")
for sensor in sensors:
    for i in range(288):  # 24 hours * 12 readings/hour (every 5 min)
        timestamp = base_time + timedelta(minutes=i * 5)
        
        # Row key: sensor_id + reverse_timestamp (for time-series ordering)
        # Reverse timestamp ensures latest data comes first in scans
        reverse_ts = str(9999999999 - int(timestamp.timestamp()))
        row_key = f"{sensor['id']}_{reverse_ts}"

        # Generate sensor readings based on type
        if sensor['type'] == 'temperature':
            value = round(random.uniform(18.0, 35.0), 2)
            unit = 'celsius'
        elif sensor['type'] == 'humidity':
            value = round(random.uniform(30.0, 90.0), 2)
            unit = 'percent'
        else:
            value = round(random.uniform(980.0, 1050.0), 2)
            unit = 'hPa'

        # Store data in HBase
        table.put(row_key.encode(), {
            b'sensor:id': sensor['id'].encode(),
            b'sensor:type': sensor['type'].encode(),
            b'sensor:location': sensor['location'].encode(),
            b'reading:value': str(value).encode(),
            b'reading:unit': unit.encode(),
            b'reading:timestamp': timestamp.strftime('%Y-%m-%d %H:%M:%S').encode(),
            b'status:battery': str(random.randint(20, 100)).encode(),
            b'status:signal': str(random.randint(60, 100)).encode(),
            b'status:alert': str(value > 32 or value < 20).encode()
        })
        total_records += 1

print(f"Total records stored: {total_records}")
print(f"Records per sensor: 288 (every 5 min for 24 hours)")

# ===== Step 4: Store using Batch (Faster method) =====
print("\nStoring additional batch data...")
batch_sensor = {'id': 'SENSOR_006', 'type': 'temperature', 'location': 'Building_D'}
batch = table.batch()

for i in range(100):
    timestamp = base_time + timedelta(minutes=i * 5)
    reverse_ts = str(9999999999 - int(timestamp.timestamp()))
    row_key = f"SENSOR_006_{reverse_ts}"
    value = round(random.uniform(20.0, 30.0), 2)
    
    batch.put(row_key.encode(), {
        b'sensor:id': b'SENSOR_006',
        b'sensor:type': b'temperature',
        b'sensor:location': b'Building_D',
        b'reading:value': str(value).encode(),
        b'reading:timestamp': timestamp.strftime('%Y-%m-%d %H:%M:%S').encode(),
        b'status:battery': str(random.randint(50, 100)).encode(),
    })

batch.send()
print("Batch insert complete! (100 records for SENSOR_006)")

# ===== Step 5: Verify Storage =====
print(f"\nTotal rows in table: ~{total_records + 100}")
print("\nSample stored record:")
for key, data in table.scan(limit=1):
    print(f"  Row Key: {key.decode()}")
    for col, val in data.items():
        print(f"    {col.decode()}: {val.decode()}")


# ============================================================
# PART 2: RETRIEVE TIME-SERIES IoT SENSOR DATA (5 Marks)
# ============================================================

print("\n" + "="*60)
print("  PART 2: RETRIEVING IoT SENSOR DATA")
print("="*60)

# ===== Step 6: Get Single Row =====
print("\n[Query 1] Get single record by row key:")
row_key = f"SENSOR_001_{str(9999999999 - int(base_time.timestamp()))}"
row = table.row(row_key.encode())
if row:
    print(f"  Row Key: {row_key}")
    for col, val in row.items():
        print(f"    {col.decode()}: {val.decode()}")

# ===== Step 7: Scan All Data for a Specific Sensor =====
print("\n[Query 2] All readings for SENSOR_001 (first 10):")
print(f"  {'TIMESTAMP':<22} {'VALUE':<10} {'BATTERY':<8}")
print(f"  {'-'*22} {'-'*10} {'-'*8}")

count = 0
for key, data in table.scan(row_prefix=b'SENSOR_001'):
    if count >= 10:
        break
    ts = data.get(b'reading:timestamp', b'N/A').decode()
    val = data.get(b'reading:value', b'N/A').decode()
    bat = data.get(b'status:battery', b'N/A').decode()
    print(f"  {ts:<22} {val:<10} {bat:<8}")
    count += 1

# ===== Step 8: Scan with Row Key Range (Time Range Query) =====
print("\n[Query 3] Time range query - SENSOR_002 (06:00 to 08:00):")

start_time = datetime(2024, 6, 15, 6, 0, 0)
end_time = datetime(2024, 6, 15, 8, 0, 0)

# Reverse timestamps for range
start_key = f"SENSOR_002_{str(9999999999 - int(end_time.timestamp()))}"
stop_key = f"SENSOR_002_{str(9999999999 - int(start_time.timestamp()))}"

print(f"  {'TIMESTAMP':<22} {'HUMIDITY %':<12}")
print(f"  {'-'*22} {'-'*12}")

for key, data in table.scan(row_start=start_key.encode(), row_stop=stop_key.encode()):
    ts = data.get(b'reading:timestamp', b'').decode()
    val = data.get(b'reading:value', b'').decode()
    print(f"  {ts:<22} {val:<12}")

# ===== Step 9: Filter by Column Value =====
print("\n[Query 4] Readings with alert=True (high/low readings):")
print(f"  {'SENSOR':<12} {'TIMESTAMP':<22} {'VALUE':<10} {'ALERT':<6}")
print(f"  {'-'*12} {'-'*22} {'-'*10} {'-'*6}")

count = 0
for key, data in table.scan():
    alert = data.get(b'status:alert', b'False').decode()
    if alert == 'True' and count < 10:
        sensor_id = data.get(b'sensor:id', b'').decode()
        ts = data.get(b'reading:timestamp', b'').decode()
        val = data.get(b'reading:value', b'').decode()
        print(f"  {sensor_id:<12} {ts:<22} {val:<10} {alert:<6}")
        count += 1

# ===== Step 10: Retrieve Data by Location =====
print("\n[Query 5] All sensors in Building_A_Floor1:")
print(f"  {'SENSOR':<12} {'TYPE':<12} {'TIMESTAMP':<22} {'VALUE':<10}")
print(f"  {'-'*12} {'-'*12} {'-'*22} {'-'*10}")

count = 0
for key, data in table.scan(row_prefix=b'SENSOR_001'):
    location = data.get(b'sensor:location', b'').decode()
    if location == 'Building_A_Floor1' and count < 10:
        sensor_id = data.get(b'sensor:id', b'').decode()
        s_type = data.get(b'sensor:type', b'').decode()
        ts = data.get(b'reading:timestamp', b'').decode()
        val = data.get(b'reading:value', b'').decode()
        print(f"  {sensor_id:<12} {s_type:<12} {ts:<22} {val:<10}")
        count += 1

# ===== Step 11: Aggregate Query — Average per Sensor =====
print("\n[Query 6] Average reading per sensor:")
print(f"  {'SENSOR':<12} {'TYPE':<12} {'AVG VALUE':<12} {'READINGS':<10}")
print(f"  {'-'*12} {'-'*12} {'-'*12} {'-'*10}")

for sensor in sensors:
    values = []
    for key, data in table.scan(row_prefix=sensor['id'].encode()):
        val = data.get(b'reading:value', b'0').decode()
        values.append(float(val))
    if values:
        avg_val = sum(values) / len(values)
        print(f"  {sensor['id']:<12} {sensor['type']:<12} {avg_val:<12.2f} {len(values):<10}")

# ===== Step 12: Retrieve Latest N Readings =====
print("\n[Query 7] Latest 5 readings for SENSOR_003:")
print(f"  {'TIMESTAMP':<22} {'PRESSURE (hPa)':<16}")
print(f"  {'-'*22} {'-'*16}")

count = 0
for key, data in table.scan(row_prefix=b'SENSOR_003'):
    if count >= 5:
        break
    ts = data.get(b'reading:timestamp', b'').decode()
    val = data.get(b'reading:value', b'').decode()
    print(f"  {ts:<22} {val:<16}")
    count += 1

# ===== Step 13: Close Connection =====
connection.close()
print("\nHBase connection closed.")

File: Big Data/Using R/HBase - Store and Retrieve time-series IoT sensor data.R
# Start HBase
start-hbase.sh

# Start Thrift server (required for Python connection)
hbase thrift start

# Start REST server (required for R connection)
hbase rest start -p 8080

# Verify HBase is running
hbase shell
> status

# Required libraries
library(httr)
library(jsonlite)

# HBase REST API (start with: hbase rest start -p 8080)
base_url <- "http://localhost:8080"

# ===== Step 1: Connect and Check Status =====
cat("Connecting to HBase REST API...\n")
response <- GET(paste0(base_url, "/status/cluster"))
cat("HBase Status:", status_code(response), "\n")

# ===== Step 2: Create Table =====
table_name <- "iot_sensor_data_r"

# Delete if exists
DELETE(paste0(base_url, "/", table_name, "/schema"))

# Create table with column families
schema <- '<?xml version="1.0" encoding="UTF-8"?>
<TableSchema name="iot_sensor_data_r">
  <ColumnSchema name="sensor"/>
  <ColumnSchema name="reading"/>
  <ColumnSchema name="status"/>
</TableSchema>'

response <- PUT(
  paste0(base_url, "/", table_name, "/schema"),
  body = schema,
  content_type("text/xml"),
  add_headers(Accept = "text/xml")
)
cat(sprintf("Table '%s' created: %d\n", table_name, status_code(response)))

# ===== Step 3: Helper Function — Encode values for HBase REST =====
# HBase REST API requires base64 encoding
encode_b64 <- function(text) {
  base64enc::base64encode(charToRaw(text))
}

decode_b64 <- function(encoded) {
  rawToChar(base64enc::base64decode(encoded))
}

# ===== Step 4: Store IoT Sensor Data =====
# install.packages("base64enc") if not installed
library(base64enc)

set.seed(42)
base_time <- as.POSIXct("2024-06-15 00:00:00")
total_records <- 0

sensors <- data.frame(
  id = c("SENSOR_001", "SENSOR_002", "SENSOR_003"),
  type = c("temperature", "humidity", "pressure"),
  location = c("Building_A", "Building_B", "Building_C"),
  stringsAsFactors = FALSE
)

cat("\nStoring IoT sensor data...\n")

for (s in 1:nrow(sensors)) {
  for (i in 1:100) {  # 100 readings per sensor (every 5 min)
    timestamp <- base_time + (i - 1) * 300  # 5 min intervals
    reverse_ts <- as.character(9999999999 - as.numeric(timestamp))
    row_key <- paste0(sensors$id[s], "_", reverse_ts)
    
    # Generate reading
    if (sensors$type[s] == "temperature") {
      value <- round(runif(1, 18, 35), 2)
    } else if (sensors$type[s] == "humidity") {
      value <- round(runif(1, 30, 90), 2)
    } else {
      value <- round(runif(1, 980, 1050), 2)
    }
    
    # Build JSON payload for HBase REST
    payload <- sprintf('{
      "Row": [{
        "key": "%s",
        "Cell": [
          {"column": "%s", "$": "%s"},
          {"column": "%s", "$": "%s"},
          {"column": "%s", "$": "%s"},
          {"column": "%s", "$": "%s"},
          {"column": "%s", "$": "%s"},
          {"column": "%s", "$": "%s"}
        ]
      }]
    }',
      encode_b64(row_key),
      encode_b64("sensor:id"), encode_b64(sensors$id[s]),
      encode_b64("sensor:type"), encode_b64(sensors$type[s]),
      encode_b64("sensor:location"), encode_b64(sensors$location[s]),
      encode_b64("reading:value"), encode_b64(as.character(value)),
      encode_b64("reading:timestamp"), encode_b64(format(timestamp, "%Y-%m-%d %H:%M:%S")),
      encode_b64("status:battery"), encode_b64(as.character(sample(20:100, 1)))
    )
    
    PUT(
      paste0(base_url, "/", table_name, "/", row_key),
      body = payload,
      content_type("application/json"),
      add_headers(Accept = "application/json")
    )
    total_records <- total_records + 1
  }
  cat(sprintf("  %s: 100 records stored\n", sensors$id[s]))
}

cat(sprintf("\nTotal records stored: %d\n", total_records))


# ============================================================
# PART 2: RETRIEVE TIME-SERIES IoT SENSOR DATA
# ============================================================

cat("\n", strrep("=", 60), "\n")
cat("  PART 2: RETRIEVING IoT SENSOR DATA\n")
cat(strrep("=", 60), "\n")

# ===== Step 5: Get Single Row =====
cat("\n[Query 1] Get single record:\n")
row_key <- paste0("SENSOR_001_", as.character(9999999999 - as.numeric(base_time)))
response <- GET(
  paste0(base_url, "/", table_name, "/", row_key),
  add_headers(Accept = "application/json")
)
if (status_code(response) == 200) {
  result <- fromJSON(content(response, "text", encoding = "UTF-8"))
  cells <- result$Row[[1]]$Cell[[1]]
  cat(sprintf("  Row Key: %s\n", row_key))
  for (i in seq_len(nrow(cells))) {
    col <- decode_b64(cells$column[i])
    val <- decode_b64(cells$`$`[i])
    cat(sprintf("    %s: %s\n", col, val))
  }
}

# ===== Step 6: Scan All Data for a Specific Sensor =====
cat("\n[Query 2] Scan SENSOR_001 (first 10 readings):\n")

# Use scanner
scanner_url <- paste0(base_url, "/", table_name, "/scanner")
scanner_body <- '{
  "startRow": "",
  "endRow": "",
  "filter": {
    "type": "PrefixFilter",
    "value": ""
  },
  "batch": 10
}'

# Simple prefix scan using row prefix endpoint
response <- GET(
  paste0(base_url, "/", table_name, "/SENSOR_001*"),
  add_headers(Accept = "application/json"),
  query = list(limit = "10")
)

if (status_code(response) == 200) {
  result <- fromJSON(content(response, "text", encoding = "UTF-8"))
  rows <- result$Row
  
  cat(sprintf("  %-22s %-10s %-8s\n", "TIMESTAMP", "VALUE", "BATTERY"))
  cat(sprintf("  %-22s %-10s %-8s\n", strrep("-", 22), strrep("-", 10), strrep("-", 8)))
  
  for (i in seq_len(min(10, length(rows)))) {
    cells <- rows[[i]]$Cell[[1]]
    ts <- ""
    val <- ""
    bat <- ""
    for (j in seq_len(nrow(cells))) {
      col <- decode_b64(cells$column[j])
      cell_val <- decode_b64(cells$`$`[j])
      if (col == "reading:timestamp") ts <- cell_val
      if (col == "reading:value") val <- cell_val
      if (col == "status:battery") bat <- cell_val
    }
    cat(sprintf("  %-22s %-10s %-8s\n", ts, val, bat))
  }
}

# ===== Step 7: Retrieve by Row Key Range (Time Range) =====
cat("\n[Query 3] Time range query - SENSOR_002 (first hour):\n")

start_time <- as.POSIXct("2024-06-15 00:00:00")
end_time <- as.POSIXct("2024-06-15 01:00:00")

start_key <- paste0("SENSOR_002_", as.character(9999999999 - as.numeric(end_time)))
stop_key <- paste0("SENSOR_002_", as.character(9999999999 - as.numeric(start_time)))

response <- GET(
  paste0(base_url, "/", table_name, "/", start_key, ",", stop_key),
  add_headers(Accept = "application/json")
)

if (status_code(response) == 200) {
  result <- fromJSON(content(response, "text", encoding = "UTF-8"))
  cat("  Time range data retrieved successfully\n")
}

# ===== Step 8: Summary Statistics =====
cat("\n[Query 4] Summary per sensor:\n")
cat(sprintf("  %-12s %-12s %-10s\n", "SENSOR", "TYPE", "RECORDS"))
cat(sprintf("  %-12s %-12s %-10s\n", strrep("-", 12), strrep("-", 12), strrep("-", 10)))

for (s in 1:nrow(sensors)) {
  cat(sprintf("  %-12s %-12s %-10d\n", sensors$id[s], sensors$type[s], 100))
}

# ===== Step 9: Get Latest Readings =====
cat("\n[Query 5] Latest 5 readings for SENSOR_003:\n")
response <- GET(
  paste0(base_url, "/", table_name, "/SENSOR_003*"),
  add_headers(Accept = "application/json"),
  query = list(limit = "5")
)

if (status_code(response) == 200) {
  result <- fromJSON(content(response, "text", encoding = "UTF-8"))
  cat(sprintf("  %-22s %-16s\n", "TIMESTAMP", "PRESSURE (hPa)"))
  cat(sprintf("  %-22s %-16s\n", strrep("-", 22), strrep("-", 16)))
  
  rows <- result$Row
  for (i in seq_len(min(5, length(rows)))) {
    cells <- rows[[i]]$Cell[[1]]
    ts <- ""
    val <- ""
    for (j in seq_len(nrow(cells))) {
      col <- decode_b64(cells$column[j])
      cell_val <- decode_b64(cells$`$`[j])
      if (col == "reading:timestamp") ts <- cell_val
      if (col == "reading:value") val <- cell_val
    }
    cat(sprintf("  %-22s %-16s\n", ts, val))
  }
}

cat("\n", strrep("=", 60), "\n")
cat("  RETRIEVAL COMPLETE!\n")
cat(strrep("=", 60), "\n")